lib/datalog/access.go

```diff
@@ -0,0 +1,380 @@
+	}
+}
+
+func (r *AccessResponse) createRoleMapping(role types.Role) {
+	r.addToMap(role.GetName())
+	for _, login := range append(role.GetLogins(types.Allow), role.GetLogins(types.Deny)...) {
+		r.addToMap(login)
+	}
+	for key, values := range role.GetNodeLabels(types.Allow) {
+		r.addToMap(key)
+		for _, value := range values {
+			r.addToMap(value)
+		}
+	}
+	for key, values := range role.GetNodeLabels(types.Deny) {
+		r.addToMap(key)
+		for _, value := range values {
+			r.addToMap(value)
+		}
+	}
+}
+
+func (r *AccessResponse) createNodeMapping(node types.Server) {
+	r.addToMap(node.GetHostname())
+	for key, value := range node.GetAllLabels() {
+		r.addToMap(key)
+		r.addToMap(value)
+	}
+}
+
+func (r *AccessResponse) createUserPredicates(user types.User, login string) {
+	r.createUserMapping(user)
+	for _, role := range user.GetRoles() {
+		r.addPredicate(hasRole, user.GetName(), role)
+	}
+	for _, trait := range user.GetTraits()[teleport.TraitLogins] {
+		r.addPredicate(hasTrait, user.GetName(), loginTraitHash, trait)
+	}
+}
+
+func (r *AccessResponse) createRolePredicates(role types.Role) {
+	r.createRoleMapping(role)
+	for _, login := range role.GetLogins(types.Allow) {
+		r.addPredicate(roleAllowsLogin, role.GetName(), login)
+	}
+	for _, login := range role.GetLogins(types.Deny) {
+		r.addPredicate(roleDeniesLogin, role.GetName(), login)
+	}
+	for key, values := range role.GetNodeLabels(types.Allow) {
+		for _, value := range values {
+			r.addPredicate(roleAllowsNodeLabel, role.GetName(), key, value)
+		}
+	}
+	for key, values := range role.GetNodeLabels(types.Deny) {
+		for _, value := range values {
+			r.addPredicate(roleDeniesNodeLabel, role.GetName(), key, value)
+		}
+	}
+}
+
+func (r *AccessResponse) createNodePredicates(node types.Server) {
+	r.createNodeMapping(node)
+	for key, value := range node.GetAllLabels() {
+		r.addPredicate(nodeHasLabel, node.GetHostname(), key, value)
+	}
+	r.addPredicate(nodeHasLabel, node.GetHostname(), types.Wildcard, types.Wildcard)
+}
+
+func (r *AccessResponse) generateAtomStrings(key string) [][]string {
+	var ret [][]string
+	for _, pred := range r.Accesses[key] {
+		var atoms []string
+		for _, atom := range pred.Atoms {
+			atoms = append(atoms, r.reverseMappings[atom])
+		}
+		ret = append(ret, atoms)
+	}
+	return ret
+}
+
+// BuildStringOutput creates the UI for displaying access responses.
+func (r *AccessResponse) BuildStringOutput() string {
+	accessTable := asciitable.MakeTable([]string{"User", "Login", "Node", "Allowing Roles"})
+	allowingRoles := make(map[string][]string)
+	accessStrings := r.generateAtomStrings(accesses)
+	for _, atoms := range accessStrings {
+		key := createDupMapKey(atoms[:roleIndex])
+		allowingRoles[key] = append(
+			allowingRoles[key],
+			atoms[roleIndex],
+		)
+	}
+	for _, key := range sortKeys(allowingRoles) {
+		sort.Strings(allowingRoles[key])
+		accessTable.AddRow(append(strings.Split(key, keyJoin), strings.Join(allowingRoles[key], ", ")))
```

Comment:
I usually try to keep the logic of converting objects from one representation into another separate from the "presentation" logic. So in my view this function does 2 things: 
1. Converts AccessResponse to some format that's more convenient to output.
2. Builds a table out of it.

I would try to split those in 2. So for example you could have a "ToTable()" method on AccessResponse that would assemble a data structure that's easy to output. And then your output logic (which can just live in the tool/ package) will just iterate over it and print the table.